### ISSUE TITLE
Feature Request: STDERR redirected through websockets

### ISSUE THREAD

**Issue by EvanCarroll** [joewalnes/websocketd#403]
It seems this program swallows stderr output,

    ./websocketd --loglevel=debug --port 8080 --devconsole -- perl -E'$|++; say for 1,2,3,4; warn $_ for 7,7,7; die 42;'

Yet when I check out `ws://localhost:8080` I don't see the messages for 7,7,7 or 42 anywhere...

```
open
ws://x1c7:8080/ 
onopen
 
onmessage
1 
onmessage
2 
onmessage
3 
onmessage
4 
onclose
[Clean: false, Code: 1006, Reason: none] 
send »
```

Is there a way to tell `websocketd` to wrap the stream in JSON and tell us what stream it is, or something of the like?

**Comment by EvanCarroll** [comment]
It seems this was brought up here https://github.com/joewalnes/websocketd/issues/315

**Comment by asergeyev** [comment]
What you're looking for goes to log (also it's not printing 7, since warn does not uses `$_` implicitly) :
![image](https://user-images.githubusercontent.com/715656/113074575-565ea880-9199-11eb-9ca0-96d1a21a178e.png)

How would you expect things to behave here?

**Comment by asergeyev** [comment]
We discussed that before besides #315. Normally you can just send your stderr to stdout and read all of the lines in Javascript.

Prefixing them or changing how they sent in the socket would go into territory of messing with program's output and it's not something websocketd should be concerned with.

**Comment by EvanCarroll** [comment]
Yes, it does go into the log which is better than nothing, but that's not really great if you're building a terminal interface to a shell program. Your users are not going to want `stderr` silently dropped or logged remotely. They're going to want it made available to the end user on web sockets.

Just consider an repl/eval bot/ or anything else. Anytime something is wrong the user who invoked the command gets nothing.

I'm not arguing against the log. I'm just saying it would be nice to not have to ad-hoc ship them back, and to have access to this stream right away.

**Comment by asergeyev** [comment]
Perhaps it reminds me #141; there could be middleware that mixes stderr/stdout

**Comment by ctn-malone** [comment]
@EvanCarroll : might not be exactly what you're looking for but I ended up working on some sort of generic wrapper which outputs 3 json-based events (stdout, stderr & exit). See https://github.com/joewalnes/websocketd/issues/402

**Comment by joewalnes** [comment]
Feature request: redirect stderr to WebSocket client. By design, stderr goes to logs for debugging. Scripts can merge stderr to stdout if needed: `2>&1`.

**Comment by joewalnes** [comment]
Oops, this issue was prematurely closed… still looking into it.

**Comment by joewalnes** [comment]
Yeah, upon reflection of this old issue, I believe the right thing to do is follow the unix semantics of stdout for program output (to the browser) and stderr for additional informational info (to the logs).

For the specific use-case of "if you're building a terminal interface to a shell program", building a custom wrapper for your own needs is the way to go, which can combine the streams and optionally wrap with some kind of app specific marker so you know which is which.

Closing this out.
